fem/kernels/diffusion.cpp

```diff
@@ -19,18 +19,162 @@ namespace kernels
 namespace fem
 {
 
-#ifdef __OCCA__
+// ****************************************************************************
+// * OCCA 2D Assemble kernel
 // *****************************************************************************
-static void oBiPADiffusionMultAdd2D(const int NUM_DOFS_1D,
-                                    const int NUM_QUAD_1D,
+#ifdef __OCCA__
+static void occaDiffusionAssemble2D(const int NUM_QUAD_1D,
                                     const int numElements,
-                                    const double* __restrict dofToQuad,
-                                    const double* __restrict dofToQuadD,
-                                    const double* __restrict quadToDof,
-                                    const double* __restrict quadToDofD,
-                                    const double* __restrict oper,
-                                    const double* __restrict solIn,
-                                    double* __restrict solOut)
+                                    const double* __restrict quadWeights,
+                                    const double* __restrict J,
+                                    const double COEFF,
+                                    double* __restrict oper)
+{
+   const int NUM_QUAD_2D = NUM_QUAD_1D*NUM_QUAD_1D;
+
+   GET_OCCA_CONST_MEMORY(quadWeights);
+   GET_OCCA_CONST_MEMORY(J);
+   GET_OCCA_MEMORY(oper);
+
+   NEW_OCCA_PROPERTY(props);
+   SET_OCCA_PROPERTY(props, NUM_QUAD_1D);
+   SET_OCCA_PROPERTY(props, NUM_QUAD_2D);
+
+   NEW_OCCA_KERNEL(Assemble2D, fem, bidiffusionAssemble.okl, props);
+   Assemble2D(numElements, o_quadWeights, o_J, COEFF, o_oper);
+}
+#endif // __OCCA__
+
+// *****************************************************************************
+// * Diffusion Assemble 2D kernel
+// *****************************************************************************
+static void DiffusionAssemble2D(const int NUM_QUAD_1D,
+                                const int numElements,
+                                const double* __restrict quadWeights,
+                                const double* __restrict J,
+                                const double COEFF,
+                                double* __restrict oper)
+{
+   GET_CONST_PTR(quadWeights);
+   GET_CONST_PTR(J);
+   GET_PTR(oper);
+   const int NUM_QUAD = NUM_QUAD_1D*NUM_QUAD_1D;
+   MFEM_FORALL(e,numElements,
+   {
+      for (int q = 0; q < NUM_QUAD; ++q)
+      {
+         const double J11 = d_J[ijklNM(0,0,q,e,2,NUM_QUAD)];
```

Comment:
The data layout represented by the `ijklNM` function will be critical for optimizing these kernels. However does the `MFEM_FORALL` will know which `ijklNM` function to use according to the device?